src/System.Private.CoreLib/src/System/Threading/ThreadPool.Unix.cs

```diff
@@ -4,18 +4,279 @@
+        /// Whether or not the wait is a repeating wait.
+        /// </summary>
+        internal bool Repeating { get; }
+
+        /// <summary>
+        /// The <see cref="WaitHandle"/> the user passed in via <see cref="Unregister(WaitHandle)"/>.
+        /// </summary>
+        private SafeWaitHandle UserUnregisterWaitHandle { get; set; } = new SafeWaitHandle((IntPtr)(-1), false); // Initialize with an invalid handle like CoreCLR
+
+        private IntPtr UserUnregisterWaitHandleValue { get; set; } = new IntPtr(-1);
+
+        internal bool IsBlocking => UserUnregisterWaitHandleValue == (IntPtr)(-1);
+
+        /// <summary>
+        /// The <see cref="ClrThreadPool.WaitThread"/> this <see cref="RegisteredWaitHandle"/> was registered on.
+        /// </summary>
+        internal ClrThreadPool.WaitThread WaitThread { get; set; }
+
+        /// <summary>
+        /// The number of callbacks that are currently queued on the Thread Pool or executing.
+        /// </summary>
+        private int _numRequestedCallbacks;
+
+        private LowLevelLock _callbackLock = new LowLevelLock();
+
+        /// <summary>
+        /// Notes if we need to signal the user's unregister event after all callbacks complete.
+        /// </summary>
+        private bool _signalAfterCallbacksComplete;
+
+        private int _unregisterCalled;
+
+        private bool _unregistered;
+
+        private AutoResetEvent _callbacksComplete;
+
+        private AutoResetEvent _removed;
+
+        /// <summary>
+        /// Unregisters this wait handle registration from the wait threads.
+        /// </summary>
+        /// <param name="waitObject">The event to signal when the handle is unregistered.</param>
+        /// <returns>If the handle was successfully marked to be removed and the provided wait handle was set as the user provided event.</returns>
+        /// <remarks>
+        /// This method will only return true on the first call.
+        /// Passing in a wait handle with a value of -1 will result in a blocking wait, where Unregister will not return until the full unregistration is completed.
+        /// </remarks>
         public bool Unregister(WaitHandle waitObject)
         {
-            // UNIXTODO: ThreadPool
-            throw new NotImplementedException();
+            if (Interlocked.Exchange(ref _unregisterCalled, 1) == 0)
+            {
+                GC.SuppressFinalize(this);
+
+                _callbackLock.Acquire();
+                try
+                {
+                    UserUnregisterWaitHandle = waitObject?.SafeWaitHandle;
+                    UserUnregisterWaitHandle?.DangerousAddRef();
+                    UserUnregisterWaitHandleValue = UserUnregisterWaitHandle?.DangerousGetHandle() ?? IntPtr.Zero;
+                    if (_unregistered)
+                    {
+                        SignalUserWaitHandle();
+                        return true;
+                    }
+
+                    _removed = new AutoResetEvent(false);
```

Comment:
The non-blocking case is a common one. Please cache an event in a static field lazily and use Interlocked.CompareExchange with null to rent the event, and Interlocked.Exchange to return the event to the cache. Allocation would only be necessary if the cache doesn't contain an event. Returning the event to the cache can be done in the two wait functions (return it to the cache, and null out the field).

Since one event is necessary at this point, both paths (blocking or non-blocking) can use the same cache.